File: src/lib/api/compare.ts
import { invoke } from "@tauri-apps/api/core";

import type { RowRecord } from "./rows";

/** 对比样本：行摘要 + 决定各分区空态文案的签名状态。 */
export interface CompareSample {
  row: RowRecord;
  hasStyleSignature: boolean;
  hasVibeSignature: boolean;
  vibeSignatureUnreadable: boolean;
}

/** 分区①②③的分页结果。 */
export interface CompareSectionPage {
  rows: RowRecord[];
  totalCount: number;
  offset: number;
  limit: number;
}

/** 分区④：同画风全部行（上限 500 张）+ 截断标记。 */
export interface CompareModelSection {
  rows: RowRecord[];
  totalCount: number;
  truncated: boolean;
}

export function getCompareSample(rowId: number): Promise<CompareSample> {
  return invoke<CompareSample>("get_compare_sample", { rowId });
}

export function queryCompareSameArtists(
  rowId: number,
  offset: number,
  limit: number,
): Promise<CompareSectionPage> {
  return invoke<CompareSectionPage>("query_compare_same_artists", {
    rowId,
    offset,
    limit,
  });
}

export function queryCompareSameVibeDiffStyle(
  rowId: number,
  offset: number,
  limit: number,
): Promise<CompareSectionPage> {
  return invoke<CompareSectionPage>("query_compare_same_vibe_diff_style", {
    rowId,
    offset,
    limit,
  });
}

export function queryCompareSameStyleDiffVibe(
  rowId: number,
  offset: number,
  limit: number,
): Promise<CompareSectionPage> {
  return invoke<CompareSectionPage>("query_compare_same_style_diff_vibe", {
    rowId,
    offset,
    limit,
  });
}

export function queryCompareSameStyleAllModels(
  rowId: number,
): Promise<CompareModelSection> {
  return invoke<CompareModelSection>("query_compare_same_style_all_models", {
    rowId,
  });
}

File: src-tauri/src/app/runtime/compare.rs
use super::{AppRuntime, AppRuntimeError};

impl AppRuntime {
    pub(crate) fn get_compare_sample(
        &self,
        row_id: i64,
    ) -> Result<crate::db::CompareSample, AppRuntimeError> {
        self.with_database(|db| db.get_compare_sample(row_id))
    }

    pub(crate) fn query_compare_same_artists(
        &self,
        row_id: i64,
        offset: u64,
        limit: u32,
    ) -> Result<crate::db::CompareSectionPage, AppRuntimeError> {
        self.with_database(|db| db.query_compare_same_artists(row_id, offset, limit))
    }

    pub(crate) fn query_compare_same_vibe_diff_style(
        &self,
        row_id: i64,
        offset: u64,
        limit: u32,
    ) -> Result<crate::db::CompareSectionPage, AppRuntimeError> {
        self.with_database(|db| db.query_compare_same_vibe_diff_style(row_id, offset, limit))
    }

    pub(crate) fn query_compare_same_style_diff_vibe(
        &self,
        row_id: i64,
        offset: u64,
        limit: u32,
    ) -> Result<crate::db::CompareSectionPage, AppRuntimeError> {
        self.with_database(|db| db.query_compare_same_style_diff_vibe(row_id, offset, limit))
    }

    pub(crate) fn query_compare_same_style_all_models(
        &self,
        row_id: i64,
    ) -> Result<crate::db::CompareModelSection, AppRuntimeError> {
        self.with_database(|db| db.query_compare_same_style_all_models(row_id))
    }
}
